### PR TITLE
Put TUN down on client disconnect.

### PR DESCRIPTION
**Pull request by joshperry**
On startup client creates TUN interface in UP state and kept it UP even
if connection to the server was lost. Creating interface in DOWN state,
turning it UP on successful (re-)connection to server and DOWN on either
disconnect or connection loss would enable DHCP client (say dhclient5)
to detect necessity for lease renewal.

Added a client configuration parameter to create TUN interface in DOWN
state and commands to enable, disable, and query the configuration
parameter.
Enabling the parameter causes client to put all unused TUN interfaces
DOWN, create new TUN interfaces in DOWN state, and turn TUN interfaces
corresponding to active sessions DOWN on connection loss or
disconnecting from server.
Disabling the parameter forces client to turn all TUN interfaces UP and
create new TUN interfaces in UP state.
Default value is 'Disable'.

### CODE REVIEW

**Comment by youxiaojie** [comment]
great

**Comment by chipitsine** [comment]
hello, @joshperry 

am I right that default behaviour is not changed by this patch ?
behaviour is only changed if explicitely enabled?

**Comment by chipitsine** [comment]
@joshperry , are you here ?

**Comment by davidebeatrici** [comment]
> Default value is 'Disable'.

File: src/Cedar/Client.h [hunk 1]
```diff
@@ -195,6 +195,7 @@ struct CLIENT_CONFIG
 	UINT KeepConnectProtocol;				// Protocol
 	UINT KeepConnectInterval;				// Interval
 	bool NoChangeWcmNetworkSettingOnWindows8;	// Don't change the WCM network settings on Windows 8
+	bool NicDownOnDisconnect;				// Put NIC down on disconnec/connection loss and put it up again after connecting to VPN server
```
Comment (davidebeatrici) [inline src/Cedar/Client.h]: Typo: `disconnec` -> `disconnect`

File: src/bin/hamcore/strtable_en.stb [hunk 2]
```diff
@@ -4664,6 +4664,21 @@ CMD_ServerCipherSet_Args	ServerCipherSet [name]
 CMD_ServerCipherSet_[name]	This specifies the encrypted and electronic signature algorithm to set. You can obtain the list of usable algorithms by using the ServerCipherGet command.
 CMD_ServerCipherSet_PROMPT_NAME	Encrypted Algorithm Name to Specify: 
 
+# TUNDownOnDisconnectEnable command
+CMD_TUNDownOnDisconnectEnable	Enable the Turn TUN Interface Down on Client Disconnect Function.
+CMD_TUNDownOnDisconnectEnable_Help	This allows you to enable Turn TUN Interface Down on Clien Disconnect Function.
```
Comment (davidebeatrici) [inline src/bin/hamcore/strtable_en.stb]: Typo: `Clien` -> `Client`

File: src/bin/hamcore/strtable_en.stb [hunk 3]
```diff
@@ -4664,6 +4664,21 @@ CMD_ServerCipherSet_Args	ServerCipherSet [name]
...
 CMD_ServerCipherSet_PROMPT_NAME	Encrypted Algorithm Name to Specify: 
 
+# TUNDownOnDisconnectEnable command
+CMD_TUNDownOnDisconnectEnable	Enable the Turn TUN Interface Down on Client Disconnect Function.
+CMD_TUNDownOnDisconnectEnable_Help	This allows you to enable Turn TUN Interface Down on Clien Disconnect Function.
+CMD_TUNDownOnDisconnectEnable_Args	TUNDownOnDisconnectEnable
+
+# TUNDownOnDisconnectDisable command
+CMD_TUNDownOnDisconnectDisable	Disable the Turn TUN Interface Down on Client Disconnect Function.
+CMD_TUNDownOnDisconnectDisable_Help	This allows you to disable Turn TUN Interface Down on Clien Disconnect Function.
```
Comment (davidebeatrici) [inline src/bin/hamcore/strtable_en.stb]: Typo: `Clien` -> `Client`

File: src/bin/hamcore/strtable_en.stb [hunk 4]
```diff
@@ -4664,6 +4664,21 @@ CMD_ServerCipherSet_Args	ServerCipherSet [name]
...
+CMD_TUNDownOnDisconnectDisable_Args	TUNDownOnDisconnectDisable
+
+# TUNDownOnDisconnectGet command
+CMD_TUNDownOnDisconnectGet		Get status of the Turn TUN Interface Down on Client Disconnect Function.
+CMD_TUNDownOnDisconnectGet_Help	This allows you to get status of the Turn TUN Interface Down on Clien Disconnect Function.
```
Comment (davidebeatrici) [inline src/bin/hamcore/strtable_en.stb]: Typo: `Clien` -> `Client`

**Comment by davidebeatrici** [comment]
Because the network interface's state should always reflect the state of the associated connection, especially because you can use it in a script to easily detect whether the VPN is connected.

**Comment by davidebeatrici** [comment]
`VLanUnix.c` is not compiled on Windows:
https://github.com/SoftEtherVPN/SoftEtherVPN/blob/30a554c1902407a7ad3118833c77604796473b17/src/Cedar/VLanUnix.c#L133

File: src/Cedar/Protocol.c [hunk 5]
```diff
@@ -5385,6 +5385,12 @@ bool ClientConnect(CONNECTION *c)
 
 	PrintStatus(sess, _UU("STATUS_9"));
```
Comment (davidebeatrici) [inline src/Cedar/Protocol.c]: `#ifdef OS_UNIX`

File: src/Cedar/Session.c [hunk 6]
```diff
@@ -1381,6 +1381,11 @@ void CleanupSession(SESSION *s)
 		FreePacketAdapter(s->PacketAdapter);
 	}
```
Comment (davidebeatrici) [inline src/Cedar/Session.c]: `#ifdef OS_UNIX`

File: src/Cedar/Session.c [hunk 7]
```diff
@@ -1529,6 +1534,11 @@ void ClientThread(THREAD *t, void *param)
 		CLog(s->Cedar->Client, "LC_CONNECT_ERROR", s->ClientOption->AccountName,
 			GetUniErrorStr(s->Err), s->Err);
```
Comment (davidebeatrici) [inline src/Cedar/Session.c]: `#ifdef OS_UNIX`

Comment (chipitsine) [reply src/Cedar/Session.c]: makes sense

File: src/Cedar/VLanUnix.c [hunk 8]
```diff
@@ -560,8 +569,10 @@ int UnixCreateTapDeviceEx(char *name, char *prefix, UCHAR *mac_address)
 		StrCpy(ifr.ifr_name, sizeof(ifr.ifr_name), macos_eth_name);
 		ioctl(s, SIOCGIFFLAGS, &ifr);
 
-		ifr.ifr_flags |= IFF_UP;
-		ioctl(s, SIOCSIFFLAGS, &ifr);
+		if (create_up) {
```
Comment (davidebeatrici) [inline src/Cedar/VLanUnix.c]: The bracket should be in a new line.

File: src/Cedar/Command.c [hunk 9]
```diff
@@ -6641,6 +6644,136 @@ UINT PcRemoteDisable(CONSOLE *c, char *cmd_name, wchar_t *str, void *param)
...
+	ret = CcGetClientConfig(pc->RemoteClient, &t);
+
+	if (ret == ERR_NO_ERROR)
+	{
+		// Change the settings
+		t.NicDownOnDisconnect = false;
+		ret = CcSetClientConfig(pc->RemoteClient, &t);
+	}
+
+	if (ret == ERR_NO_ERROR)
```
Comment (davidebeatrici) [inline src/Cedar/Command.c]: To be removed.

File: src/Cedar/Command.c [hunk 10]
```diff
@@ -6641,6 +6644,136 @@ UINT PcRemoteDisable(CONSOLE *c, char *cmd_name, wchar_t *str, void *param)
...
+	ret = CcGetClientConfig(pc->RemoteClient, &t);
+
+	if (ret == ERR_NO_ERROR)
+	{
+		// Change the settings
+		t.NicDownOnDisconnect = true;
+		ret = CcSetClientConfig(pc->RemoteClient, &t);
+	}
+
+	if (ret == ERR_NO_ERROR)
```
Comment (davidebeatrici) [inline src/Cedar/Command.c]: To be removed.

**Comment by joshperry** [comment]
Hey guys,

The importance of this patch is like @davidebeatrici said, but not just for connectivity detection. All network management scripts on Linux rely on the transitioning of the Up/Down state to know when certain configuration tasks should be done.

For example the dhcpcd client only attempts to allocate/renew a dynamic IP address when the interface transitions from Down to Up.

Another critical aspect is that the Kernel will keep this network segment as part of the system's routing table as long as it is Up. This makes it very difficult to implement features like least-cost or priority routing based on the network metrics if the network is not really Up when it says it is.

Thank you for pushing this forward!

Josh

**Comment by joshperry** [comment]
@dav

**Comment by joshperry** [comment]
@davidebeatrici Did you accidentally overwrite this PR?

**Comment by davidebeatrici** [comment]
@joshperry Yes, because I pushed before completing the rebase and thus I left out 190ad01eb758f85ae20b945cfc75c7204da88d84.

I completed the rebase, force-pushed and 59e1483dbf707b1adcec04f5f450c0b3445ab5f2 was deleted.

**Comment by davidebeatrici** [comment]
Thank you very much for your contribution!

Given the importance of the network interface's state, I think that we should enable `NicDownOnDisconnect` by default in future.

**Comment by joshperry** [comment]
Excellent, I didn't realize you were working on it. Thank you!

I agree on the default change as well.

**Pull request merged**